docs/Operations.md

```diff
@@ -33,35 +58,144 @@ Maestro will have multiple operations, and those will be set using pairs of defi
+    canceled(Canceled)
+    error(Error)
+    
+    pending --> in_progress;
+    pending --> evicted;
+    in_progress --> finished;
+    in_progress --> error;
+    in_progress --> canceled;
+	
+```
+
 ## Lifecycle
+```mermaid
+flowchart TD
+    finish((End))
+    
+    created("Created (Pending)")
+    evicted(Evicted)
+    error(Error)
+    finished(Finished)
+    canceled(Canceled)
+    
+    should_execute{Should Execute?}
+    execution_succeeded{Success?}
+
+    execute[[Execute]]
+    rollback[[Rollback]]
+
+    canceled_by_user>Canceled By User]
+    
+    created --> should_execute;
+    should_execute -- No --> evicted --> finish;
+    should_execute -- Yes --> execute;
+    execute --> execution_succeeded;
+    execute --> canceled_by_user --> canceled --> rollback;
+    execution_succeeded -- Yes --> finished --> finish;
+    execution_succeeded -- No --> error --> rollback --> finish;
 ```
-                     ┌────────────────────┐
-                     │Create new operation│
-                     └──────────┬─────────┘
-                                │
-                                │
-                         ┌──────▼──────┐
-                         │Pending      │
-                         └──────┬──────┘
-                                │
-                                │
-                 ┌──────────────▼────────────────┐
-                 │Will the operation be executed?│
-                 └──────┬─────────────────┬──────┘
-                        │Yes              │No
-                        │                 │
-                 ┌──────▼──────┐   ┌──────▼──────┐
-                 │In progress  │   │Evicted      │
-                 └──────┬──────┘   └─────────────┘
-        Succeeds        │
-        ┌───────────────┤
-        │               │Fails
-        │               │
- ┌──────▼──────┐ ┌──────▼──────┐
- │Finished     │ │ Error       │
- └─────────────┘ └─────────────┘
+
+## Lease
+### What is the operation lease
+Lease is a mechanism to track operations execution and help troubleshooting long/delayed operations. 
+### Why Operations have it
+Sometimes, an operation might get stuck. It could happen, for example, if the worker crashes during the execution of an operation.
+To keep track of operations, we assign each operation a Lease.
+This Lease has a TTL (time to live). 
+
+When the operation is being executed, this TTL is renewed each time the lease is about to expire but the operation is still in progress.
+It'll be revoked once the operation is finished.
+
+### Troubleshooting
+If an operation is fetched and the TTL expired (the TTL is in the past), the operation probably got stuck.
+
+If an operation does not have a Lease, it either did not start at all (should be on the queue) or is already finished. 
+An Active Operation without a Lease is at an invalid state.
+
+
+>**⚠ Maestro do not have a self-healing routine yet for expired operations.**
+
+
+### Operation Lease Lifecycle
+```mermaid
+flowchart TD
```

Comment:
The operation rollback is executed before changing the state to `error` or `rollback`